pallets/pallet-did-lookup/src/migrations.rs

```diff
@@ -16,74 +16,151 @@
+
+// Inspired by frame_support::storage::migration::move_storage_from_pallet
+fn move_storage<P: PalletInfoAccess>(old_storage_name: &[u8], new_storage_name: &[u8]) {
+	let pallet_name = <P as PalletInfoAccess>::name();
+
+	let old_prefix = storage_prefix(pallet_name.as_bytes(), old_storage_name);
+	let new_prefix = storage_prefix(pallet_name.as_bytes(), new_storage_name);
+	move_prefix(&old_prefix, &new_prefix);
+
+	if let Some(value) = unhashed::get_raw(&old_prefix) {
+		unhashed::put_raw(&new_prefix, &value);
+		unhashed::kill(&old_prefix);
 	}
+}
 
+pub struct EthereumMigration<T>(PhantomData<T>);
+
+impl<T: crate::pallet::Config> OnRuntimeUpgrade for EthereumMigration<T>
+where
+	T::AccountId: Into<LinkableAccountId>,
+{
 	fn on_runtime_upgrade() -> frame_support::weights::Weight {
-		// Account for the new storage version written below.
-		let initial_weight = T::DbWeight::get().writes(1);
-
-		// Origin was disabled, so there cannot be any existing links. But we check just
-		// to be sure.
-		let total_weight: Weight =
-			ConnectedDids::<T>::iter().fold(initial_weight, |total_weight, (account, record)| {
-				ConnectedAccounts::<T>::insert(record.did, account, ());
-				// One read for the `ConnectedDids` entry, one write for the new
-				// `ConnectedAccounts` entry.
-				total_weight.saturating_add(T::DbWeight::get().reads_writes(1, 1))
+		if Pallet::<T>::current_storage_version() == StorageVersion::new(3) {
+			// already on version 3
+			<T as frame_system::Config>::DbWeight::get().reads_writes(1, 0)
+		} else {
+			log::info!("🔎 DidLookup: Initiating migration");
+			let mut connected_dids = 0u64;
+			let mut connected_accounts = 0u64;
+
+			// Migrate connected DIDs
+			// We should not write to the same storage item during drain because it can lead
+			// to undefined results. Thus, we write to a temporary storage and move that at
+			// the end. Else we iterate over every key more or less twice.
+			ConnectedDids::<T>::drain().for_each(|(acc_id32, value)| {
+				log::debug!(
+					"🔎 #{:?} Migrating ConnectedDid for account id {:?}",
+					connected_dids.encode(),
+					acc_id32.encode()
+				);
+				let acc_id: LinkableAccountId = acc_id32.into();
+				TmpConnectedDids::<T>::insert(acc_id, value);
+				connected_dids += 1;
+			});
+			log::info!("🔎 DidLookup: Migrated all {:?} ConnectedDids", connected_dids);
+
+			// Migrate accounts
+			ConnectedAccounts::<T>::drain().for_each(|(did_id, acc_id32, val)| {
+				log::debug!(
+					"🔎 #{:?} Migrating ConnectedAccount for did_id {:?} and account id {:?}",
+					connected_accounts,
+					did_id.encode(),
+					acc_id32.encode()
+				);
+				let acc_id: LinkableAccountId = acc_id32.into();
+				TmpConnectedAccounts::<T>::insert(did_id, acc_id, val);
+				connected_accounts += 1;
 			});
+			log::info!("🔎 DidLookup: Migrated all {:?} ConnectedAccounts", connected_accounts);
+
+			// Move TmpStorage
+			move_storage::<Pallet<T>>(b"TmpConnectedDids", b"ConnectedDids");
```

Comment:
Use rather the `move_storage::Pallet::<T>` syntax.

Reply:
doesn't work, can you suggest how its better?

Reply:
Sorry you are right. I mistook this for something else. We decided a while ago as a styling convention to prefer `Type::<T>::operation()` over `<Type<T>>::operation()` since they are both valid syntaxes. But this is a different case, so I will resolve the comment, sorry for the spam 😄